File: C++/zoo.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

class Animal {
public:
    virtual string info() = 0;
};

class Cat: public Animal {
public:
    virtual string info() {
        return "Cat";
    }
};

class Dog: public Animal {
public:
    virtual string info() {
        return "Dog";
    }
};

int main() {
    vector<Animal*> zoo {new Cat(), new Cat(), new Dog(), new Dog()};
    for (auto x : zoo) cout << x->info() << "\n";
}

File: OCaml/zoo.ml
class animal = object
  method info = "Animal"  
end

class dog init = object
  inherit animal
  method info = "Dog"
end

class cat init = object
  inherit animal as super
  method info = "Cat is an " ^ super#info
end

let zoo = [new dog(); new cat()];;

List.iter (fun x -> print_endline x#info) zoo